### PR TITLE
Allow using non default validator for FormContext.

### LINKED ISSUE #16349: FormValidation with custom validator are not applied on Form
### Description

I defined a form and add the option 'context' => ['validator' => 'private'] to define that form follow a custom validator. I defined in my customerModel a validatorDefault and a validatorPrivate($validator). The validatorPrivate call the validatorDefault and extend the ruleset. The rule with regex are not applied in the form, when i put the rule in validatorPrivate. But when I put the rule in teh validatorDefault the rule will be applied and the error is shown in the form. Rule with notEmptSting will be apllay in both validators. All rules that use the $validator->add will be only apllied in validatorDefault.

```
snipped validator of the customerTable.php

/**
   * Validator for private validation
   *
   * @param \Cake\Validation\Validator $validator the validator
   * @return \Cake\Validation\Validator
   */
  public function validationPrivate(Validator $validator): Validator
  {
      $validator = $this->validationDefault($validator);

      $validator
          ->notEmptyString('Name_1', 'Bitte geben Sie einen Vorname ein.')
          ->regex('Name_1', $this->pattern, 'Vorname enthält ungültige Zeichen');
      $validator
          ->notEmptyString('Name_2', 'Bitte geben Sie einen Nachname ein.')
          ->regex('Name_2', $this->pattern, 'Nachname enthält ungültige Zeichen');

      return $validator;
  }
```

regex rule will not applay in my form.

### CakePHP Version

4.3.5

### PHP Version

_No response_

### PR DESCRIPTION
**Pull request by ADmad**
Closes #16349

### CODE REVIEW

File: src/View/Form/FormContext.php [hunk 1]
```diff
@@ -33,17 +35,31 @@ class FormContext implements ContextInterface
...
+
     /**
      * Constructor.
      *
      * @param array $context Context info.
+     *
+     * Keys:
+     *
+     * - `entity` The Form class instance this context is operating on.
+     * - `validator` Optional name of the validation method to call on the Form object.
```
Comment (othercorey) [inline src/View/Form/FormContext.php:53]: We need a better format for these in general, but I don't have a suggestion right now.

Comment (othercorey) [inline src/View/Form/FormContext.php:53]: We need a better format for these in general, but I don't have a suggestion right now.

File: src/View/Form/FormContext.php [hunk 2]
```diff
@@ -33,17 +35,31 @@ class FormContext implements ContextInterface
...
+    protected $_validator = null;
+
     /**
      * Constructor.
      *
      * @param array $context Context info.
+     *
+     * Keys:
+     *
+     * - `entity` The Form class instance this context is operating on.
```
Comment (othercorey) [inline src/View/Form/FormContext.php]:
```suggestion
     * - `entity` The Form class instance this context is operating on. **(required)**
```
```suggestion
     * - `entity` The Form class instance this context is operating on.
```

Comment (othercorey) [inline src/View/Form/FormContext.php]:
```suggestion
     * - `entity` The Form class instance this context is operating on. **(required)**
```
```suggestion
     * - `entity` The Form class instance this context is operating on.
```

File: src/View/Form/FormContext.php [hunk 3]
```diff
@@ -33,17 +35,31 @@ class FormContext implements ContextInterface
...
+     * - `validator` Optional name of the validation method to call on the Form object.
      */
     public function __construct(array $context)
     {
-        $context += [
-            'entity' => null,
```
Comment (othercorey) [inline src/View/Form/FormContext.php:44]: Is removing this default a BC break or does it already fail further into processing?

Comment (othercorey) [inline src/View/Form/FormContext.php:44]: Is removing this default a BC break or does it already fail further into processing?

Comment (ADmad) [reply src/View/Form/FormContext.php:44]: It's not a BC break, even currently you would have got an error trying to use any of its methods if entity wasn't set. I added an explicit check and exception now.

File: src/View/Form/FormContext.php (hunk 1 above)
Comment (ADmad) [reply src/View/Form/FormContext.php:53]: Yeah the "entity" could be the 1st arg which can be be typed and other optional config in 2nd array argument. Something we could do in 5.x.

**Comment by othercorey** [comment]
Looks ok to me, but I'm not that familiar with FormContext.

File: src/View/Form/FormContext.php [hunk 4]
```diff
@@ -33,17 +35,31 @@ class FormContext implements ContextInterface
      */
     protected $_form;
 
+    /**
+     * Validator name.
+     *
+     * @var string|null
+     */
+    protected $_validator = null;
```
Comment (markstory) [inline src/View/Form/FormContext.php:43]: Do we need the `_` prefix on new properties?

Comment (ADmad) [reply src/View/Form/FormContext.php:43]: Not really, I was just being consistent with the name used in EntityContext.

**Review by markstory — APPROVED** [review]

**Pull request merged**